libevmasm/Assembly.h

```diff
@@ -65,8 +65,50 @@ class Assembly
 	}
 
 	std::optional<uint8_t> eofVersion() const { return m_eofVersion; }
+	bool supportsFunctions() const { return m_eofVersion.has_value(); }
+	bool supportsRelativeJumps() const { return m_eofVersion.has_value(); }
 	AssemblyItem newTag() { assertThrow(m_usedTags < 0xffffffff, AssemblyException, ""); return AssemblyItem(Tag, m_usedTags++); }
 	AssemblyItem newPushTag() { assertThrow(m_usedTags < 0xffffffff, AssemblyException, ""); return AssemblyItem(PushTag, m_usedTags++); }
+	AssemblyItem newFunctionCall(uint16_t _functionID)
+	{
+		solAssert(_functionID < m_codeSections.size(), "Call to undeclared function.");
+		solAssert(_functionID > 0, "Cannot call call section 0");
+		auto const& section = m_codeSections.at(_functionID);
+		if (section.outputs != 0x80)
+			return AssemblyItem::functionCall(_functionID, section.inputs, section.outputs);
+		else
+			return AssemblyItem::jumpF(_functionID, section.inputs);
+	}
+
+	AssemblyItem newFunctionReturn()
+	{
+		return AssemblyItem::functionReturn(m_codeSections.at(m_currentCodeSection).outputs);
+	}
+
+	uint16_t createFunction(uint8_t _args, uint8_t _rets)
+	{
+		size_t functionID = m_codeSections.size();
+		solAssert(functionID < 1024, "Too many functions.");
+		solAssert(m_currentCodeSection == 0, "Functions need to be declared from the main block.");
+		solAssert(_rets <= 0x80, "Too many function returns.");
+		m_codeSections.emplace_back(CodeSection{_args, _rets, {}});
+		return static_cast<uint16_t>(functionID);
+	}
+
+	void beginFunction(uint16_t _functionID)
+	{
+		solAssert(m_currentCodeSection == 0, "Atempted to begin a function before ending the last one.");
+		solAssert(_functionID < m_codeSections.size(), "Attempt to begin an undeclared function.");
+		auto& section = m_codeSections.at(_functionID);
+		solAssert(section.items.empty(), "Function already defined.");
+		m_currentCodeSection = _functionID;
+	}
+	void endFunction()
+	{
+		solAssert(m_currentCodeSection != 0, "End function without begin function.");
+		m_currentCodeSection = 0;
+	}
+
```

Comment:
These are long enough that I'd rather put them in the .cpp file.

Reply:
Right. Fixed.